### PR TITLE
Simplified `HTTPClient.Completion` and refactored response handling implementation

### PR DESCRIPTION
**Pull request by NachoSoto**
For #695.

The biggest change is that this code at the beginning of `HTTPClient.handle` is now gone:
```swift
var statusCode: HTTPStatusCode = .networkConnectTimeoutError
var jsonObject: [String: Any]?
var httpResponse: HTTPResponse? = HTTPResponse(statusCode: statusCode, jsonObject: jsonObject)
var receivedJSONError: Error?
```

It was hard to trace which paths modified these default values. The new implementation has a mostly linear path (part of it abstracted in smaller methods now) and doesn’t have mutable values.

This is the new `handle` implementation:
```swift
func handle(urlResponse: URLResponse?,
            request: Request,
            urlRequest: URLRequest,
            data: Data?,
            error networkError: Error?) {
    let (response, shouldRetry) = self.parse(
        urlResponse: urlResponse,
        request: request,
        urlRequest: urlRequest,
        data: data,
        error: networkError
    )

    if shouldRetry {
        Logger.debug(Strings.network.retrying_request(httpMethod: request.method.httpMethod,
                                                      path: request.path))

        self.state.modify {
            $0.queuedRequests.insert(request.retriedRequest(), at: 0)
        }
    } else {
        request.completionHandler?(response)
    }

    self.beginNextRequest()
}
```

## All Changes:

- `HTTPClient.perform` no longer has a bunch of mutable variables with forking paths that make it hard to follow. Most of the response handling is now in a new `parse` method. This will simplify the task of continuing to refactor response deserialization.
- `HTTPClient`’s completion abstracts response into `HTTPResponse`
- `UserInfoAttributeParser` is now an `enum`
- `ETagManager` no longer takes an `Error`. The only thing it did with it is to return the default response.
- `HTTPResponse.jsonObject` is no longer `Optional`
- Moved `defaultJsonEncoder` / `defaultJsonDecoder` to `JSONDecoder+Extensions`. They used by `HTTPClient` and others soon.

### CODE REVIEW

File: Sources/Networking/HTTPClient.swift [hunk 1]
```diff
@@ -160,62 +162,112 @@ private extension HTTPClient {
...
-                    } catch let jsonError {
-                        Logger.error(Strings.network.parsing_json_error(error: jsonError))
-
-                        let dataAsString = String(data: data, encoding: .utf8) ?? ""
-                        Logger.error(Strings.network.json_data_received(dataString: dataAsString))
-
-                        receivedJSONError = jsonError
-                    }
+    // swiftlint:disable:next function_body_length
+    func parse(urlResponse: URLResponse?,
```
Comment (NachoSoto) [inline Sources/Networking/HTTPClient.swift:166]: These changes are hard to follow in the diff, but I did this very slowly thinking about every possible path. Also our test coverage gives me a lot of confidence these are equivalent.

File: Tests/UnitTests/Networking/ETagManagerTests.swift [hunk 2]
```diff
@@ -72,31 +71,12 @@ class ETagManagerTests: XCTestCase {
...
         let httpURLResponse = httpURLResponseForTest(url: url, eTag: eTag, statusCode: .success)
         let response = eTagManager.httpResultFromCacheOrBackend(
-            with: httpURLResponse, jsonObject: responseObject, error: nil, request: request, retried: false)
+            with: httpURLResponse, jsonObject: responseObject, request: request, retried: false)
         expect(response).toNot(beNil())
         expect(response?.statusCode) == .success
         expect(response?.jsonObject as? [String: String]).to(equal(responseObject))
     }
 
-    func testBackendResponseIsReturnedIfThereIsAnError() {
```
Comment (NachoSoto) [inline Tests/UnitTests/Networking/ETagManagerTests.swift:81]: `ETagManager` no longer takes an `error`, so this test isn't relevant anymore. Instead `HTTPClient` skips calling here since it was a no-op.

File: Tests/UnitTests/Networking/HTTPClientTests.swift [hunk 3]
```diff
@@ -289,13 +285,13 @@ class HTTPClientTests: XCTestCase {
...
-            message.value = result.value?["message"] as? String
+        self.client.perform(request, authHeaders: [:]) {
+            result.value = $0
         }
 
-        expect(message.value).toEventually(equal("something is great up in the cloud"), timeout: .seconds(1))
-        expect(isResponseCorrect.value) == true
+        expect(result.value).toEventuallyNot(beNil(), timeout: .seconds(1))
+        expect(result.value).to(beSuccess())
+        expect(result.value?.value?.jsonObject["message"] as? String) == "something is great up in the cloud"
```
Comment (NachoSoto) [inline Tests/UnitTests/Networking/HTTPClientTests.swift:294]: A future PR will make `HTTPClient` not return `.success` when requests fail. This doesn't change the test, but soon this will become `expect(result.value).to(beFailure())` instead.

Comment (NachoSoto) [reply Tests/UnitTests/Networking/HTTPClientTests.swift:294]: #1431

File: Sources/Networking/HTTPClient.swift [hunk 4]
```diff
@@ -160,62 +162,112 @@ private extension HTTPClient {
...
+                                                                     retried: request.retried)
+            }
+
+        let shouldRetry: Bool = {
+            if case .success(.none) = result {
+                return true
+            } else {
+                return false
+            }
+        }()
```
Comment (aboedo) [inline Sources/Networking/HTTPClient.swift:237]: it took me a bit to figure out that this means that we should retry if we got an empty but successful result. 
I think this is equivalent, right? Using our Result extensions
```swift
let shouldRetry = result.error == nil && result.value == nil
```

File: Sources/Networking/Operations/Handling/CustomerInfoResponseHandler.swift [hunk 5]
```diff
@@ -15,15 +15,10 @@ import Foundation
 
 class CustomerInfoResponseHandler {
 
-    let userInfoAttributeParser: UserInfoAttributeParser
-
-    init(userInfoAttributeParser: UserInfoAttributeParser = UserInfoAttributeParser()) {
-        self.userInfoAttributeParser = userInfoAttributeParser
-    }
+    init() { }
```
Comment (aboedo) [inline Sources/Networking/Operations/Handling/CustomerInfoResponseHandler.swift:18]: are we not injecting this for tests? how are we testing this class?

File: Sources/FoundationExtensions/JSONDecoder+Extensions.swift [hunk 6]
```diff
@@ -129,3 +129,17 @@ extension Encodable {
...
 }
+
+let defaultJsonEncoder: JSONEncoder = {
+    let encoder = JSONEncoder()
+    encoder.keyEncodingStrategy = .convertToSnakeCase
+
+    return encoder
+}()
+
+let defaultJsonDecoder: JSONDecoder = {
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/FoundationExtensions/JSONDecoder+Extensions.swift:140]: Much better!! I had a TODO earlier to figure out a better place for this but couldn't think of one. That's much better 👍🏻

File: Sources/Networking/HTTPClient.swift [hunk 7]
```diff
@@ -160,62 +162,112 @@ private extension HTTPClient {
...
+                        } else {
+                            return error
+                        }
+                    },
+                shouldRetry: false
+            )
+        }
+
+        guard let httpURLResponse = urlResponse as? HTTPURLResponse else {
+            return (.failure(ErrorUtils.unexpectedBackendResponseError()), false)
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Networking/HTTPClient.swift:189]: I actually changed this approach in #1431 so it's short lived.

File: Sources/Networking/HTTPClient.swift [hunk 8]
```diff
@@ -160,62 +162,112 @@ private extension HTTPClient {
...
-                    self.state.modify {
-                        $0.queuedRequests.insert(request.retriedRequest(), at: 0)
-                    }
+                guard let result = result else {
+                    throw ErrorUtils.unexpectedBackendResponseError()
                 }
+
+                return result
+            } else {
+                return [:]
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Networking/HTTPClient.swift:213]: 👍🏻  I'll add a comment

File: Sources/Networking/HTTPClient.swift [hunk 9]
```diff
@@ -160,62 +162,112 @@ private extension HTTPClient {
...
+                    },
+                shouldRetry: false
+            )
+        }
+
+        guard let httpURLResponse = urlResponse as? HTTPURLResponse else {
+            return (.failure(ErrorUtils.unexpectedBackendResponseError()), false)
+        }
+
+        func extractBody(_ response: HTTPURLResponse, _ statusCode: HTTPStatusCode) throws -> [String: Any] {
```
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Networking/HTTPClient.swift:192]: Mainly because this is only used by `parse`. But might be more readable to take it out I agree.

File: Sources/Networking/HTTPClient.swift (hunk 4 above)
Comment (NachoSoto) [reply Sources/Networking/HTTPClient.swift:237]: I agree, I hated this too. I replaced how this is done in #1431 with a `Result` extension.

File: Sources/Networking/Operations/Handling/CustomerInfoResponseHandler.swift (hunk 5 above)
Comment (NachoSoto) [reply Sources/Networking/Operations/Handling/CustomerInfoResponseHandler.swift:18]: It wasn't being mocked, so instead of injecting it this PR turned it into an `enum`. But `UserInfoAttributeParser` actually goes away in #1431.

File: Sources/Networking/HTTPClient.swift (hunk 9 above)
[the remark this answers is not in the document]
Comment (NachoSoto) [reply Sources/Networking/HTTPClient.swift:192]: Do you mind if I do that after #1431 to simplify merge conflicts?

**Comment by NachoSoto** [comment]
Could I get a review on this one? It's the main change that all the other PRs are based on 🙏🏻

File: Sources/Networking/HTTPClient.swift (hunk 9 above)
[the remark this answers is not in the document]
Comment (aboedo) [reply Sources/Networking/HTTPClient.swift:192]: sure! 👍

**Review by aboedo — APPROVED** [review]
